### ISSUE TITLE
IWYU suggest forward declarations when it should not

### ISSUE THREAD

**Issue by DoDoENT** [include-what-you-use/include-what-you-use#375]
I've attached a little [project](https://github.com/include-what-you-use/include-what-you-use/files/601378/forward_declare.zip) which demonstrates the issue.

There are 3 classes: A, B and StringSerializer. A is data holder, B can serialize A to any `std::ostream` and StringSerializer uses B to serialize A to `std::string`.
B's `serialize` method use default parameter for `A`, therefore B's header **requires** A's header to be included - IWYU suggest forward declaring class which of course causes compile error.
Next, A has method `parse_from_string` which accepts `std::string` by reference with default string. IWYU erroneously suggest changing include of `<string>` to `<iosfwd>` which causes compile error (for instatiation of default parameter, full type information is required).

The `StringSerializer` class intended to demonstrate another issue I have on my real project - when `std::stringstream` is used in inline implementation of method in header, IWYU sometimes suggest replacing `<sstream>` with `<ostream>`, which causes compile errors. Unfortunately, this simple project does not reproduce this issue (I will try to make a minimum reproducable example later).

Here is the full build output:

```
11:37:26: Starting: "/Applications/CMake.app/Contents/bin/cmake" --build . --target all
[1/6 0.0/sec] Building CXX object CMakeFiles/forward_declare-lib.dir/B.cpp.o
Warning: include-what-you-use reported diagnostics:

/Users/dodo/Desktop/iwyu_report/case1/forward_declare/B.hpp should add these lines:
class A;

/Users/dodo/Desktop/iwyu_report/case1/forward_declare/B.hpp should remove these lines:
- #include "A.hpp"  // lines 4-4

The full include-list for /Users/dodo/Desktop/iwyu_report/case1/forward_declare/B.hpp:
#include <iosfwd>  // for ostream
class A;
---

/Users/dodo/Desktop/iwyu_report/case1/forward_declare/B.cpp should add these lines:
#include "A.hpp"    // for A

/Users/dodo/Desktop/iwyu_report/case1/forward_declare/B.cpp should remove these lines:

The full include-list for /Users/dodo/Desktop/iwyu_report/case1/forward_declare/B.cpp:
#include "B.hpp"
#include <ostream>  // for basic_ostream, operator<<, endl
#include "A.hpp"    // for A
---

[2/6 0.0/sec] Building CXX object CMakeFiles/forward_declare-lib.dir/StringSerializer.cpp.o
[3/6 0.0/sec] Building CXX object CMakeFiles/forward_declare-lib.dir/A.cpp.o
Warning: include-what-you-use reported diagnostics:

/Users/dodo/Desktop/iwyu_report/case1/forward_declare/A.hpp should add these lines:
#include <iosfwd>  // for string

/Users/dodo/Desktop/iwyu_report/case1/forward_declare/A.hpp should remove these lines:
- #include <string>  // lines 3-3

The full include-list for /Users/dodo/Desktop/iwyu_report/case1/forward_declare/A.hpp:
#include <iosfwd>  // for string
---

(/Users/dodo/Desktop/iwyu_report/case1/forward_declare/A.cpp has correct #includes/fwd-decls)

[4/6 0.0/sec] Linking CXX static library libforward_declare-lib.a
[5/6 0.0/sec] Building CXX object CMakeFiles/forward_declare.dir/main.cpp.o
[6/6 0.0/sec] Linking CXX executable forward_declare
11:37:28: The process "/Applications/CMake.app/Contents/bin/cmake" exited normally.
11:37:28: Elapsed time: 00:02.
```

I am using MacOS 10.12, Clang from XCode 8.1 (clang-800.0.42.1), QtCreator 4.1 with CMake 3.7 and Ninja 1.7.1 and IWYU 0.7 downloaded from [this link](http://include-what-you-use.org/downloads/include-what-you-use-0.7-x86_64-apple-darwin.tar.gz).

Besides that, IWYU also suggested using some private headers on my system, so I would like to share my mapping that so far suppressed those behaviour (I will update this mapping as I find more problems):

```
[
    { include: ["@<opencv2/.*>", private, "<opencv2/opencv.hpp>", public] },
    { include: ["@<.*_types.*>", private, "<cstdint>", public] },
    { include: ["<__tree>", private, "<utility>", public] },
    { include: ["<__functional_base>", private, "<functional>", public] },
    { include: ["@<.*_pthread.*>", private, "<pthread.h>", public] },
    { include: ["\"gtest/gtest.h\"", public, "<gtest/gtest.h>", public] },
    { include: ["<math.h>", public, "<cmath>", public] },
    { include: ["<__bit_reference>", private, "<iterator>", public] },
    { include: ["<stddef.h>", public, "<cstddef>", public] },
    { include: ["<ostream>", public, "<sstream>", public] },
    { include: ["<sys/dirent.h>", private, "<dirent.h>", public] },
    { include: ["<sys/errno.h>", private, "<errno.h>", public] }
]
```

As you can see, mapping already has `<ostream>` mapped to `<sstream>`, which works for my code, but is not optimal.

**Comment by kimgr** [comment]
> IWYU erroneously suggest changing include of `<string>` to `<iosfwd>`
> which causes compile error (for instatiation of default parameter, full type information is required).

This is a problem with IWYU's understanding of libc++, it thinks string is provided by `<iosfwd>`, which is nonsense, of course. I've seen the stream problems too (we have a bug for a couple of strange scenarios).

**Comment by kimgr** [comment]
> B's serialize method use default parameter for A, therefore B's header requires A's header to be included > - IWYU suggest forward declaring class which of course causes compile error.

It looks like default arguments are just not handled by IWYU, so this would be nice to get in place.

**Comment by kimgr** [comment]
The `<iosfwd>`/`<string>` portion of this bug was silently fixed in https://github.com/include-what-you-use/include-what-you-use/commit/bc5221b570f4c70eb2601a79c79da2832a1c7aac, released as part of IWYU 0.21.

Thanks @bolshakov-a for taking care of this!

@DoDoENT Please reopen if you still have issues around default arguments.
